src/commands/SinceCommand.ts

```diff
@@ -0,0 +1,240 @@
+            return { ok: { minDate: maybeMinDate, maxAgeMS: maybeMaxAgeMS} };
+        }
+
+        //...or as a duration
+        maybeMaxAgeMS = parseDuration(source);
+        if (maybeMaxAgeMS && !Number.isNaN(maybeMaxAgeMS)) {
+            maybeMaxAgeMS = Math.abs(maybeMaxAgeMS);
+            return { ok: { minDate: new Date(Date.now() - maybeMaxAgeMS), maxAgeMS: maybeMaxAgeMS } }
+        }
+        return { error: "" };
+    });
+    if ("error" in minDateResult) {
+        return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", minDateResult.error);
+    }
+    const { minDate, maxAgeMS } = minDateResult.ok!;
+
+    // Parse max entries.
+    const maxEntriesResult = parseToken("<maxEntries>", maxEntriesToken, source => {
+        const maybeMaxEntries = Number.parseInt(source, 10);
+        if (Number.isNaN(maybeMaxEntries)) {
+            return { error: "Not a number" };
+        } else {
+            return { ok: maybeMaxEntries };
+        }
+    });
+    if ("error" in maxEntriesResult) {
+        return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", maxEntriesResult.error);
+    }
+    const maxEntries = maxEntriesResult.ok!;
+
+    // Attempt to parse `<action>` as Action.
+    const actionResult = parseToken("<action>", actionToken, source => {
+        for (let key in Action) {
+            const maybeAction = Action[key as keyof typeof Action];
+            if (key === source) {
+                return { ok: maybeAction }
+            } else if (maybeAction === source) {
+                return { ok: maybeAction }
+            }
+        }
+        return {error: `Expected one of ${JSON.stringify(Action)}`};
+    })
+    if ("error" in actionResult) {
+        return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", actionResult.error);
+    }
+    const action: Action = actionResult.ok!;
+
+    // Now list affected rooms.
+    const rooms: Set<string> = new Set();
+    let reason: string | undefined;
+    for (let i = 0; i < optionalTokens.length; ++i) {
+        const token = optionalTokens[i];
+        const maybeRoomResult = getTokenAsString("[room]", token);
+        if ("error" in maybeRoomResult) {
+            return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", maybeRoomResult.error);
+        }
+
+        const maybeRoom = maybeRoomResult.ok!;
+
+        if (maybeRoom === "*") {
+            for (let roomId of Object.keys(mjolnir.protectedRooms)) {
+                rooms.add(roomId);
+            }
+        } else if (maybeRoom.startsWith("#") || maybeRoom.startsWith("!")) {
+            const roomId = await mjolnir.client.resolveRoom(maybeRoom);
+            if (!(roomId in mjolnir.protectedRooms)) {
+                return mjolnir.logMessage(LogLevel.WARN, "SinceCommand", `This room is not protected: ${htmlEscape(roomId)}.`);
+            }
+            rooms.add(roomId);
+        } else {
+            if (i === 0) {
+                // First argument may not be a room.
```

Comment:
if we switched around the order of `reason` and `rooms`, you can pop off args while they start with `#` or `!` and collect everything else as a reason

Reply:
Done.